scene/main/scene_tree.cpp

```diff
@@ -1217,6 +1217,38 @@ void SceneTree::get_nodes_in_group(const StringName &p_group, List<Node *> *p_li
 void SceneTree::_flush_delete_queue() {
 	_THREAD_SAFE_METHOD_
 
+	// This section is an optimization because nodes benefit immensely from being deleted
+	// in reverse order to their child count. This is partly due to ordered_remove(), and partly
+	// due to notifications being sent to children that are moved, further in the child list.
+
+	// Less need to apply this when deleting small numbers of nodes.
+	// This magic number is guessed .. to get most of the benefit while not checking through
+	// child lists when unnecessary (as this will have a cost).
+	if (delete_queue.size() > 128) {
```

Comment:
Is there performance degradation for <128 nodes?

Reply:
There is a cost to the technique. If a parent has e.g. 1000 children, and only one is queued for deletion, then it will iterate through all the children _just in case_ any of them are queued. This could worst case result in a bunch of cache misses. So we ideally don't want to use it _unless_ we know it will help.

We know there is no possible benefit in using the optimization with a queue size of 1, and we know the benefits tend to scale approximately with `number of children being deleted / number of children` (on average, it depends on the exact deletion patterns), so it makes sense to turn it on at some overall queue size between 2 and 1000 (which was measured somewhat inaccurately as 1.3x faster earlier).

In the same way there is less absolute benefit possible in a parent that has few children, as the effect scales with the log of the number of children.

The magic numbers in both cases could probably be decreased (maybe as a result of further testing), but this conservative approach should be good to start because it will eliminate most long pauses.

There are a few more complex techniques which could be used to achieve the same thing in a less "haphazard" fashion, such as sorting the queue delete list, or having a list of queue deleted nodes on each node, but the main difficulty is that currently each node has no idea of its child ID relative to its parent. This could be fixed by storing a child ID per node and keeping it up to date as children are added / moved, but it is pretty easy for bugs to creep in here and for it to get out of sync. So the simple approach in this PR is probably more sensible to get started (unless profiling shows the same thing is a problem in other situations).

Another variation is we could store a counter of the number of queue deleted children per node, and use this as a heuristic "number in queue / number of children" to turn on or off.

Reply:
I understand where you're coming from. It's definitely a band-aid for a larger problem regarding deletion performance. It makes more sense to look deeper into the deletion queue for a cleaner solution. 

I'm newer to the engine so I apologize if this is off. 

But it appears `node`'s know their [depth](https://github.com/godotengine/godot/blob/706dc74e01e1900af947d7a5fe854603a3df18b1/scene/main/node.h#L108) at the time they are [removed](https://github.com/godotengine/godot/blob/706dc74e01e1900af947d7a5fe854603a3df18b1/scene/main/scene_tree.cpp#L1032). Maybe having different queues for each `scene_tree` level would be more efficient? That way you'll always delete the deepest children first.

Reply:
This seems to be a misunderstanding of the problem. The relevant factor is the child ID (i.e. which number child in the list of the parent), the depth is unrelated to the problem (afaik).

Reply:
Maybe? The PR appears to be looking for children of the parent to delete first, because deleting the parent before the children is causing the issue. 

Wouldn't the depth determine what would potentially be a child? IE, the deepest node would always be a child. 

So there would be no need to check all of a parent's children to see if they need to be deleted first, as anything deeper than the parent(all of its potential children) was already deleted from the deletion queue.

Reply:
> Maybe? The PR appears to be looking for children of the parent to delete first, because deleting the parent before the children is causing the issue.
> 
> Wouldn't the depth determine what would potentially be a child? IE, the deepest node would always be a child.
> 
> So there would be no need to check all of a parent's children to see if they need to be deleted first, as anything deeper than the parent(all of its potential children) was already deleted from the deletion queue.

You seem to be misunderstanding the issue and the PR, all of your suppositions here are incorrect. I can only suggest re-reading the issue, and examining what the MRP does.